### ISSUE TITLE
Incorrect behavior of `AddMethodCallBasedStrictParamTypeRector`

### ISSUE THREAD

**Issue by mvhirsch** [rectorphp/rector#8507]
# Bug Report

I :heart: rector. It made my dreams come true. Keep on the good work!

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | 0.19.8      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.com/demo/fbcd0e42-42ad-4aa8-b35c-89c7233d4ab1

```php
<?php

final class DemoFile
{
    private function get(): ?\DateTime
    {
        return null;
    }
    
    public function run1(bool $param)
    {
        return $this->format($this->get());
    }
                             
    public function run2(bool $param)
    {
        return $this->format(new \DateTimeImmutable());
    }

    private function format(\DateTimeInterface $date)
    {
        return $date->format('Y-m-d');
    }
}
```

### Responsible rules

* `AddMethodCallBasedStrictParamTypeRector`

## Expected Behavior

While the code is broken (I must admit that), because `null->format()` won't work at all, Rector should not change it `\DateTimeInterface` to `null|\DateTime|\DateTimeImmutable`, but to `?\DateTimeInterface` as both implement that interface and it was prior defined.

So basically I expected something like this:

```diff
-    private function format(\DateTimeInterface $date)
+    private function format(?\DateTimeInterface $date)
     {
         return $date->format('Y-m-d');
     }
```

**Comment by samsonasik** [comment]
I think that's trade off based on caller, check instanceof inside union type each other sub type is memory intensive, as it will make double loop per-type, and code transformation is valid. I think you can verify code after transformation by hand ;)

**Comment by mvhirsch** [comment]
Yeah, I thought about that, too. For me it was a tricky one: is it a bug, or a feature? Guess it's the latter :joy:

**Comment by samsonasik** [comment]
That's a chore I guess
